File: practices/exercism/ruby/2019/collatz_conjecture.rb
# Collatz Conjecture from https://exercism.io

class CollatzConjecture
  class << self
    def steps(number, count = 0)
      raise ArgumentError if number <= 0

      number == 1 ? count : steps(calculate_with(number), count += 1)
    end

    private

      def calculate_with(number)
        number.odd? ? number * 3 + 1 : number / 2
      end
  end
end

File: sandbox/exercism/go/2025/collatz_conjecture.go
// https://exercism.org/tracks/go/exercises/collatz-conjecture

package collatzconjecture

import "errors"

func CollatzConjecture(n int) (int, error) {
	if 0 >= n { return 0, errors.New("invalid number") }

	step := 0

	for {
		if n == 1 {
			return step, nil
		} else if n % 2 == 0 {
			n = n / 2
		} else {
			n = (n * 3 + 1)
		}

		step++
	}
}
